Article web-pwa-native (#323)
oddbird/oddleventy: content/blog/2023/when-to-choose-a-responsive-web-app.md

---
title: When to Choose a Responsive Web App
sub: Part 1 - Responsive Web App vs PWA vs Native Mobile App
author: sondra
date: 2023-05-05
tags:
  - Article
  - Business Development
  - Push Notifications
  - Recommendations
  - Startups
  - User Experience
image:
  src: blog/2023/part1-web.jpg
  alt: |
    spider web with dew drops

> summary: |

Reviewer: There's a lot of duplication between the summary and the opening paragraph. Since we show both in order on the page, that stands out.

You have an idea for a custom digital product you’d like to build, and it’s time to decide how to build it. One of your friends swears you won’t succeed unless you build a native mobile app for iPhone. A colleague argues that a responsive web app is the only way to go. Who’s right? Are there alternative options? The decision of how to build your app is quite important – affecting project scope, timeline, and budget – but understanding when to choose one option over another can be overwhelming.

The great news is that you don’t need to spend years becoming a web developer yourself in order to make a wise choice. In the following three-part series, we’ll dig into the three main options for building mobile apps: Responsive Web Apps, Progressive Web Apps, and Native Mobile Apps, and how to determine the best option for your project. As you read, think about who your app will serve, what their most pressing needs are, and how your app will meet those needs.

**Read all three parts to compare app types:**

1. [When to Choose a Responsive Web App](/2023/05/05/when-to-choose-a-responsive-web-app/)
2. [When to Choose a Native Mobile App](/2023/06/07/when-to-choose-a-native-mobile-app/)

> 3. [When to Choose a Progressive Web App](/2023/07/05/when-to-choose-a-progressive-web-app/)

Reviewer: If we're publishing these a month apart, does one work well before the others are available?

Reviewer 2: Related: If we want to commit the second two in this PR without publishing them, we should add:

```yml
eleventyExcludeFromCollections: true
permalink: false
```

Reviewer 2: Added these, but commented out for easier review.

## What are Responsive Web Apps, PWAs, and Native Mobile Apps?
Here are the biggest differences in a nutshell.

> - **Responsive Web App (RWA)** - An application that runs in a browser, such as Chrome, Firefox, or Safari, and responds to fit any screen size on any device.

Reviewer: ```suggestion
- **Responsive Web App (RWA)** - An application that runs in web browsers, such as Chrome, Firefox, or Safari, and responds to fit any screen size on any device.
```

Don't want to imply that a good responsive app might only run in a single browser (which is common, and not great)

> - **Progressive Web App (PWA)** - A responsive web app with additional capabilities that mimic a native mobile app.

Reviewer: It might help to have some examples here, like opening them as their own app (rather than a browser tab) or using them while offline?

- **Native Mobile App** - An app built specifically for one particular mobile operating system, like iOS or Android.

{{ embed.img(
  src='blog/2023/comparison-chart.jpg',
  alt='A comparison chart showing the differences between
  responsive web apps, progressive web apps, and native
  mobile apps'

> ) }}

Reviewer: This leaves all the details of the chart pretty inaccessible. Could we do the comparison in HTML, rather than in an image?

(also a little skeptical about 'best capabilities' as a heading. Most maybe? Or most device-specific capabilities?)

Author: @mirisuzanne @stacyk Stacy and I discussed releasing this comparison chart as an image temporarily and then building it in HTML. This article is intended to help promote newsletter sign up, so we didn't want to have to delay publication while waiting for the chart to be built out.

Author: @mirisuzanne @stacyk 
I added a summary of the comparison chart. I don't assume I've done it in the right place, but the content is here: /rwa-pwa-native-comparison-chart-summary/ however you all recommend.

## Responsive Web Apps - Best for Wide Reach

> **TL;DR**
>
> If your top priority is to reach as wide a range of
> people as possible, as quickly as possible, while keeping
> development and maintenance costs down, a responsive web app is
> likely the best option for building your digital product.

An RWA is essentially a website that resizes and rearranges its content – responding to fit whatever screen size a person chooses to use. RWAs are built using standard web technologies like HTML, CSS, and JavaScript. They are accessible from any device with a web browser such as Chrome, Firefox, Microsoft Edge, Safari, or a mobile browser.

{{ embed.figure(
  data=[{
    img: 'blog/2023/yes.png',
    alt: 'The word yes with a smiley and sparkles'
  }],
  attrs={
    'style': '--column: content-start'
  }
) }}

> ## Choose an RWA when…

Reviewer: Might be worth mentioning that an RWA can easily _become_ a PWA. So it can be a good first step, when you plan to add more functionality later.

### Lower Cost
*…when it’s important to keep initial development costs low.*

Responsive web apps use a single codebase. The term *codebase* is developer speak for the collection of code used to build an application. Compared to native mobile apps, RWAs are relatively inexpensive to develop, requiring just one codebase to be available on any device with a web browser.

To be clear, building any kind of app is not a low cost endeavor. Even a very simple responsive web app can have a starting price of $10K, and very complex web apps, such as social media platforms or project management software, cost many millions of dollars to design, develop, and maintain.

### Faster Development
*…when it’s a priority to make your app available to users quickly.*

RWAs take less time to build, relative to native mobile apps; it only takes one codebase to reach any device with a browser. It can also take less time to make the minimum viable product (MVP) available to an initial set of users, get feedback, and release frequent, incremental updates.

> Another caveat. No matter the type, even a very simple MVP of your app will take at least a few months to build. Very complex apps often take multiple years to design and develop. If you want to make sure you're meeting the needs of the people who use your app, you will want to allow time to collect their feedback along the way and make improvements.

Reviewer: ```suggestion
No matter the type, even a very simple MVP of your app will take at least a few months to build. Very complex apps often take multiple years to design and develop. If you want to make sure you're meeting the needs of the people who use your app, you will want to allow time to collect their feedback along the way and make improvements.
```

### Less Maintenance
*…when you need to keep ongoing maintenance costs low.*

RWAs require less time and expense to maintain compared to native mobile apps. An RWA’s single codebase is built with standard web technologies, and doesn’t require an update of the entire app to make incremental improvements. It also takes less time and fewer developers to update a single codebase than to update separate codebases that use specialized technology for multiple platforms.

### Discoverable Via Search Engine
*…when the people you want to reach will look for your content with a search engine.*

Search engines, like Google, crawl the web looking for keywords and other information to sort and rank the content. Then they collect, process, and store a database of all that content. This is called *indexing*. RWAs live on the web, so unless special code is added to prevent it, they will automatically be indexed and appear in organic search results.

### Broad Reach
*…when you want to reach a broad audience across devices.*

> People can access the same RWA on any device from a small iPhone to a tablet, from a laptop to a giant Windows desktop, unlike native mobile apps which are limited to supported mobile devices. Additionally, the web provides a high level of backward compatibility, meaning RWAs can be available in some form even on old devices. On the other hand, one platform update can totally block a device from accessing a native mobile app.

Reviewer: ```suggestion
People can access the same RWA on any device from a small phone to a tablet, from a laptop to a giant desktop, unlike native mobile apps which are limited to supported specific mobile devices. Additionally, the web provides a high level of backward compatibility, meaning RWAs can be available in some form even on old devices. On the other hand, one platform update can totally block a device from accessing a native mobile app.
```

{{ embed.figure(
  data=[{
    img: 'blog/2023/no.png',
    alt: 'The word no with a frowning face'
  }],
  attrs={
    'style': '--column: content-start'
  }
) }}

## Don’t choose an RWA when…
### Limited Capabilities
*…when your project requires a high level of specialized capability.*

RWAs cannot access all of a device’s hardware and software features. For example, if your digital product requires direct integration with a specific mobile device’s GPS for geofencing or access to light and proximity sensors, then an RWA won’t work.

> It is *possible* to access some advanced device features in the context of a RWA, but then you would be moving into the [progressive web app](/2023/07/05/when-to-choose-a-progressive-web-app/) space, not strictly an RWA which provides static content for people to consume.

Reviewer: > which provides static content for people to consume.

That seems overly narrow. Most of the apps we build for clients use the RWA approach, and almost none of them are 'static content for people to consume'. You don't need access to device features in order to have dynamic content.

### Performance Limitations
*…when you need advanced performance.*

If a person has a slow internet connection, they will have trouble accessing RWA content that requires a lot of processing power – like videos or animated games. If they have no internet connection, they will not be able to access an RWA at all.

### Security Limitations
*…when you require advanced security.*

If your project needs advanced security beyond typical HTTPS encryption, such as with banking applications, an RWA is less likely able to meet those requirements.

### Not Discoverable Via App Store
*…when the people you want to reach will look for your app in an app store.*

RWAs are not listed in app stores. If the people you want to reach exclusively use app stores to search for apps, they will not be able to find a responsive web app.

## In Summary

> A responsive web app tends to be less expensive to build than other types of apps -- especially the native mobile app -- and offers the broadest reach. But if have specialized needs around performance, security, or the app stores, one of the other two types of apps -- a native mobile app or a progressive web app -- are probably going to be a better fit. Let's take a look at the pros and cons of the native mobile app next.

Reviewer: > Let's take a look…

That active phrasing makes it sound like we're going right into it, but the reality is we're waiting a month.

Reviewer 3: The last line of both page 1 and 2 in the summary can probably be commented out until those posts are live IMO

> A responsive web app tends to be less expensive to build than other types of apps -- especially the native mobile app -- and offers the broadest reach. But if have specialized needs around performance, security, or the app stores, one of the other two types of apps -- a native mobile app or a progressive web app -- are probably going to be a better fit. Let's take a look at the pros and cons of the native mobile app next.

Reviewer 4: ```suggestion
A responsive web app tends to be less expensive to build than other types of apps -- especially the native mobile app -- and offers the broadest reach. But if you have specialized needs around performance, security, or the app stores, one of the other two types of apps -- a native mobile app or a progressive web app -- are probably going to be a better fit. Let's take a look at the pros and cons of the native mobile app next.
```

{{ embed.img(
  src='blog/2023/comparison-chart.jpg',
  alt='A comparison chart showing the differences between
    responsive web apps, progressive web apps, and native
    mobile apps'
) }}

## Responsive Web Apps -- Best for Wide Reach

> **TL;DR**
>
> If your top priority is to reach as wide a range of
> people as possible, as quickly as possible, while keeping
> development and maintenance costs down, a responsive web app is
> likely the best option for building your digital product.

An RWA is essentially a website that resizes and rearranges its content --
responding to fit whatever screen size a person chooses to use. RWAs are built
using standard web technologies like HTML, CSS, and JavaScript. They are
accessible from any device with a web browser such as Chrome, Firefox, Microsoft
Edge, Safari, or a mobile browser.

{{ embed.figure(
  data=[{
    img: 'blog/2023/yes.png',
    alt: 'The word "yes" with a smiley and sparkles'
  }],
  attrs={
    'style': '--column: content-start'
  }
) }}

## Choose an RWA when…

### Lower Cost
*…when it's important to keep initial development costs low.*

Responsive web apps use a single codebase. The term *codebase* is

> developer-speak for the collection of code used to build an application.

Author: I think a footnote could be a good idea. I'd actually like to go in the direction of defining more terms because I'd like people who are unfamiliar with apps to feel very comfortable reading this article.

> developer-speak for the collection of code used to build an application.

Reviewer 2: @SondraE That makes sense to me. You can look at Mia's recent `zero-units` article for how to add footnotes: https://github.com/oddbird/oddleventy/blob/main/content/blog/2022/zero-units.md?plain=1#L174-L185

> developer-speak for the collection of code used to build an application.

Reviewer 2: I moved this sentence into a footnote: https://github.com/oddbird/oddleventy/commit/23eec2e9cd9ac4bd36504274c3d6fe4b6e2bfb30

what their most pressing needs are, and how your app will meet those needs.

<!-- **Read all three parts to compare app types:**

1. [When to Choose a Responsive Web App](/2023/05/05/when-to-choose-a-responsive-web-app/)
2. [When to Choose a Native Mobile App](/2023/06/07/when-to-choose-a-native-mobile-app/)
3. [When to Choose a Progressive Web App](/2023/07/05/when-to-choose-a-progressive-web-app/) -->

## What are Responsive Web Apps, Native Mobile Apps, and Progressive Web Apps?

Here are the biggest differences in a nutshell.

- **Responsive Web App (RWA)** -- An application that runs in web browsers, such
  as Chrome, Firefox, or Safari, and responds to fit any screen size on any
  device
- **Native Mobile App** -- An app built specifically for one particular mobile
  operating system, like iOS or Android, and downloaded from an app store
- **Progressive Web App (PWA)** -- A responsive web app with additional
  capabilities that mimic a native mobile app, such as push notifications and
  offline mode

{{ embed.img(
  src='blog/2023/app-comparison-chart.jpg',
  alt='A comparison chart showing the differences between
    responsive web apps, progressive web apps, and native

>     mobile apps [summary](/rwa-pwa-native-comparison-chart-summary/)'

Reviewer 2: It won't work to embed a link like this in `alt` text, so we'll need to find another way to surface this.

---
title: When to Choose a Responsive Web App
sub: Part 1 -- Responsive Web App vs Native Mobile App vs Progressive Web App
author: sondra

> date: 2023-07-03

Reviewer 2: ```suggestion
date: 2024-07-03
```

1. [When to Choose a Responsive Web App](/2023/05/05/when-to-choose-a-responsive-web-app/)
2. [When to Choose a Native Mobile App](/2023/06/07/when-to-choose-a-native-mobile-app/)
3. [When to Choose a Progressive Web App](/2023/07/05/when-to-choose-a-progressive-web-app/) -->

## What are Responsive Web Apps, Native Mobile Apps, and Progressive Web Apps?

Here are the biggest differences in a nutshell.

- **Responsive Web App (RWA)** -- An application that runs in web browsers, such
  as Chrome, Firefox, or Safari, and responds to fit any screen size on any
  device
- **Native Mobile App** -- An app built specifically for one particular mobile
  operating system, like iOS or Android, and downloaded from an app store
- **Progressive Web App (PWA)** -- A responsive web app with additional
  capabilities that mimic a native mobile app, such as push notifications and
  offline mode

{{ embed.img(
  src='blog/2023/app-comparison-chart.jpg',
  alt='A comparison chart showing the differences between responsive web apps,
  progressive web apps, and native mobile apps'
) }}

<details>
<summary>View the text alternative of the comparison chart</summary>

> 

Reviewer 3: @mirisuzanne @jgerigmeyer is this an accessible way to have the alt text displayed for screen readers and anyone who wants selectable text? 

https://deploy-preview-323--oddleventy.netlify.app/2023/05/05/when-to-choose-a-responsive-web-app/#what-are-responsive-web-apps%2C-native-mobile-apps%2C-and-progressive-web-apps%3F

Reviewer: I think so. In practice it confused me a bit, because the headlines stay at the same level and have very similar names. Which makes me wonder if all the details of the chart are or can be worked smoothly into the content? And then the entire article is a text alternative, and the image can be labeled as more of a 'handout' version of the article?

Reviewer 2: Proposal:

- Remove summary/details, quick while @mirisuzanne is on vacation.

[^codebase]: The term *codebase* is developer-speak for the collection of code
  used to build an application.

{{ embed.img(
  src='blog/2023/htmlcssjs.jpg',

>   alt=''

Reviewer 2: Needs alt text?

Author: Added
